### PR TITLE
Contract tests for all SIDAM endpoints used by CCD

### PR DESCRIPTION
**Pull request by mohammedanas**
https://tools.hmcts.net/jira/browse/RDM-5177
https://tools.hmcts.net/jira/browse/RDM-5061

1- Initiate a new, dedicated repo for consumer driven contracts of ccd for sidam.
2- Checkout and copy-paste the content of the sample project here: https://github.com/hmcts/pact-sample-consumer-java
3- Then, change the initial content to make it work a simple, minimal suite for sidam.
4- Generate and publish the contract with Gradle tasks/commands.
5- Make sure the contract itself is visible on the Broker.
6- Get contracts verified by SIDAM.
7- Make sure the contract verification is visible on the Broker.
There maybe obvious steps missed in between numbers above.
AND
1- Find out all the endpoints of SIDAM that we're calling from any CCD application. (This can be achieved looking up idam keyword in the codes.)
2- For every such user-case, analyse what is being sent to SIDAM and what different responses are being received (happy / unhappy paths).
3- For every different use-case found above, add contract-driven test to the test suit.

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM hmcts/cnp-java-base:openjdk-8u191-jre-alpine3.9-2.0.1
```
Comment (MSancaktutar) [inline Dockerfile]: is this file really needed? don't think so.

File: bin/init.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env sh
```
Comment (MSancaktutar) [inline bin/init.sh]: bin folder and its content really needed?

File: bin/run-in-docker.sh [hunk 3]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env sh
```
Comment (MSancaktutar) [inline bin/run-in-docker.sh]: needed really?

File: charts/rpe-spring-boot-template/.helmignore [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# Patterns to ignore when building packages.
```
Comment (MSancaktutar) [inline charts/rpe-spring-boot-template/.helmignore]: charts folder and its content really needed?

File: config/owasp/suppressions.xml [hunk 5]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (MSancaktutar) [inline config/owasp/suppressions.xml]: this file is containing only the needed suppressions? if it is a copy from some other repo, its content may not be fully fit for purpose here.

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+version: '2.1'
```
Comment (MSancaktutar) [inline docker-compose.yml]: needed?

File: infrastructure/aat.tfvars [hunk 7]
```diff
@@ -0,0 +1 @@
+capacity = "2"
```
Comment (MSancaktutar) [inline infrastructure/aat.tfvars]: Are infrastructure folder and its contents really needed?

File: target/pacts/ccd_data_store-sidam_authorise_user_endpoint.json [hunk 8]
```diff
@@ -0,0 +1,70 @@
+{
```
Comment (MSancaktutar) [inline target/pacts/ccd_data_store-sidam_authorise_user_endpoint.json:1]: this file can already be generated from what we are keeping in the code. so, I believe it could be better to exclude this from version control. (adding target folder to the .gitignore file)

File: target/pacts/ccd_data_store-sidam_token.json [hunk 9]
```diff
@@ -0,0 +1,73 @@
+{
```
Comment (MSancaktutar) [inline target/pacts/ccd_data_store-sidam_token.json:1]: this file can already be generated from what we are keeping in the code. so, I believe it could be better to exclude this from version control. (adding target folder to the .gitignore file)

File: target/pacts/ccd_data_store-sidam_user_details.json [hunk 10]
```diff
@@ -0,0 +1,87 @@
+{
```
Comment (MSancaktutar) [inline target/pacts/ccd_data_store-sidam_user_details.json:1]: this file can already be generated from what we are keeping in the code. so, I believe it could be better to exclude this from version control. (adding target folder to the .gitignore file)

File: web.config [hunk 11]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (MSancaktutar) [inline web.config]: not sure if we really need this file. worth confirming.

**Review by MSancaktutar — CHANGES_REQUESTED** [review]

File: bin/init.sh (hunk 2 above)
Comment (mohammedanas) [reply bin/init.sh]: Removed

File: bin/run-in-docker.sh (hunk 3 above)
Comment (mohammedanas) [reply bin/run-in-docker.sh]: Removed

File: charts/rpe-spring-boot-template/.helmignore (hunk 4 above)
Comment (mohammedanas) [reply charts/rpe-spring-boot-template/.helmignore]: Removed

File: config/owasp/suppressions.xml (hunk 5 above)
Comment (mohammedanas) [reply config/owasp/suppressions.xml]: Removed

File: docker-compose.yml (hunk 6 above)
Comment (mohammedanas) [reply docker-compose.yml]: Removed

File: Dockerfile (hunk 1 above)
Comment (mohammedanas) [reply Dockerfile]: Removed

File: target/pacts/ccd_data_store-sidam_authorise_user_endpoint.json (hunk 8 above)
Comment (mohammedanas) [reply target/pacts/ccd_data_store-sidam_authorise_user_endpoint.json:1]: It would be nice to have the files to know what each version would generate as this is what we shall be deploying per commit

File: target/pacts/ccd_data_store-sidam_token.json (hunk 9 above)
Comment (mohammedanas) [reply target/pacts/ccd_data_store-sidam_token.json:1]: it would be nice to have the files to know what each version would generate as this is what we shall be deploying per commit

File: target/pacts/ccd_data_store-sidam_user_details.json (hunk 10 above)
Comment (mohammedanas) [reply target/pacts/ccd_data_store-sidam_user_details.json:1]: it would be nice to have the files to know what each version would generate as this is what we shall be deploying per commit

File: web.config (hunk 11 above)
Comment (mohammedanas) [reply web.config]: removed

**Comment by MSancaktutar** [comment]
Could you please also complete the README file to a minimally useful content?

File: target/pacts/ccd_data_store-sidam_user_details.json (hunk 10 above)
Comment (MSancaktutar) [reply target/pacts/ccd_data_store-sidam_user_details.json:1]: ok.

File: target/pacts/ccd_data_store-sidam_token.json (hunk 9 above)
Comment (MSancaktutar) [reply target/pacts/ccd_data_store-sidam_token.json:1]: ok.

File: target/pacts/ccd_data_store-sidam_authorise_user_endpoint.json (hunk 8 above)
Comment (MSancaktutar) [reply target/pacts/ccd_data_store-sidam_authorise_user_endpoint.json:1]: ok.

File: Dockerfile (hunk 1 above)
Comment (MSancaktutar) [reply Dockerfile]: ok.

File: bin/init.sh (hunk 2 above)
Comment (MSancaktutar) [reply bin/init.sh]: ok.

File: charts/rpe-spring-boot-template/.helmignore (hunk 4 above)
Comment (MSancaktutar) [reply charts/rpe-spring-boot-template/.helmignore]: ok.

File: config/owasp/suppressions.xml (hunk 5 above)
Comment (MSancaktutar) [reply config/owasp/suppressions.xml]: ok.

File: bin/run-in-docker.sh (hunk 3 above)
Comment (MSancaktutar) [reply bin/run-in-docker.sh]: ok.

File: Jenkinsfile_nightly [hunk 12]
```diff
@@ -0,0 +1,16 @@
+#!groovy
```
Comment (MSancaktutar) [inline Jenkinsfile_nightly]: Like the previous ones, is this file needed?

File: src/main/java/uk/gov/hmcts/sidam/IdObject.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
+package uk.gov.hmcts.sidam;
+
+public class IdObject {
+
+	private long id;
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/sidam/IdObject.java]: spaces does not look stylistically correct. pls double-check.

File: src/main/java/uk/gov/hmcts/sidam/SIDAMApi.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+import feign.Body;
+import feign.Headers;
+import feign.Param;
+import feign.RequestLine;
+import org.springframework.cloud.openfeign.FeignClient;
+import org.springframework.web.bind.annotation.*;
+
+@FeignClient(name = "SIDAMService")
+public interface SIDAMApi {
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/sidam/SIDAMApi.java]: Names could be more readible as SidamApi, SidamService, etc.

File: src/test/java/uk/gov/hmcts/sidam/SIDAMPACTTest.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+import static org.assertj.core.api.Assertions.*;
+
+@ExtendWith(PactConsumerTestExt.class)
+@ExtendWith(SpringExtension.class)
+@PactTestFor(providerName = "SIDAMService", port = "8888")
+@SpringBootTest({
+    // overriding provider address
+    "SIDAMService.ribbon.listOfServers: localhost:8888"
+})
+public class SIDAMPACTTest {
```
Comment (MSancaktutar) [inline src/test/java/uk/gov/hmcts/sidam/SIDAMPACTTest.java]: SidamPactTest - much more readible

**Review by MSancaktutar — CHANGES_REQUESTED** [review]
all good to the best of my understanding - except for the last comments to be done needful of.

File: Jenkinsfile_nightly (hunk 12 above)
Comment (mohammedanas) [reply Jenkinsfile_nightly]: removed

File: src/main/java/uk/gov/hmcts/sidam/IdObject.java (hunk 13 above)
Comment (mohammedanas) [reply src/main/java/uk/gov/hmcts/sidam/IdObject.java]: removed

**Review by MSancaktutar — APPROVED** [review]
looks ok.
